src/Lean/Data/Json/Parser.lean

```diff
@@ -44,31 +43,46 @@ def escapedChar : Parser Char := do
 
 partial def strCore (acc : String) : Parser String := do
   let c ← peek!
-  if c = '"' then -- "
+  if c == '"' then
     skip
     return acc
   else
     let c ← any
-    if c = '\\' then
+    if c == '\\' then
       strCore (acc.push (← escapedChar))
     -- as to whether c.val > 0xffff should be split up and encoded with multiple \u,
     -- the JSON standard is not definite: both directly printing the character
     -- and encoding it with multiple \u is allowed. we choose the former.
-    else if 0x0020 ≤ c.val ∧ c.val ≤ 0x10ffff then
+    else if 0x0020 <= c.val && c.val <= 0x10ffff then
       strCore (acc.push c)
     else
       fail "unexpected character in string"
 
-def str : Parser String := strCore ""
+@[inline] def str : Parser String := strCore ""
 
-partial def natCore (acc digits : Nat) : Parser (Nat × Nat) := do
-  let some c ← peek? | return (acc, digits)
-  if '0' ≤ c ∧ c ≤ '9' then
-    skip
-    let acc' := 10*acc + (c.val.toNat - '0'.val.toNat)
-    natCore acc' (digits+1)
+partial def natCore (acc : Nat) : Parser Nat := do
+  let c ← peekD '\x00'
```

Comment:
We might be able to get rid of this null byte hack with checking `eof?` and then peeking? I'd hope that the compiler is smart enough to make this not make a difference.

Reply:
Good idea. I've changed `eof?` to `isEof` to reflect our usual naming conventions and used that instead.